sway/desktop/transaction.c

```diff
@@ -0,0 +1,398 @@
+#define _POSIX_C_SOURCE 200809L
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <wlr/types/wlr_buffer.h>
+#include <wlr/types/wlr_linux_dmabuf.h>
+#include "sway/debug.h"
+#include "sway/desktop/transaction.h"
+#include "sway/output.h"
+#include "sway/tree/container.h"
+#include "sway/tree/view.h"
+#include "sway/tree/workspace.h"
+#include "list.h"
+#include "log.h"
+
+/**
+ * How long we should wait for views to respond to the configure before giving
+ * up and applying the transaction anyway.
+ */
+#define TIMEOUT_MS 200
+
+/**
+ * If enabled, sway will always wait for the transaction timeout before
+ * applying it, rather than applying it when the views are ready. This allows us
+ * to observe the rendered state while a transaction is in progress.
+ */
+#define TRANSACTION_DEBUG false
+
+struct sway_transaction {
+	struct wl_event_source *timer;
+	list_t *instructions;   // struct sway_transaction_instruction *
+	list_t *damage;         // struct wlr_box *
+	size_t num_waiting;
+	size_t num_configures;
+	struct sway_transaction *next;
+	struct timespec create_time;
+	struct timespec commit_time;
+};
+
+struct sway_transaction_instruction {
+	struct sway_transaction *transaction;
+	struct sway_container *container;
+	struct sway_container_state state;
+	struct wlr_buffer *saved_buffer;
+	uint32_t serial;
+	bool ready;
+};
+
+struct sway_transaction *transaction_create() {
+	struct sway_transaction *transaction =
+		calloc(1, sizeof(struct sway_transaction));
+	transaction->instructions = create_list();
+	transaction->damage = create_list();
+	if (server.debug_txn_timings) {
+		clock_gettime(CLOCK_MONOTONIC, &transaction->create_time);
+	}
+	return transaction;
+}
+
+static void remove_saved_view_buffer(
+		struct sway_transaction_instruction *instruction) {
+	if (instruction->saved_buffer) {
+		wlr_buffer_unref(instruction->saved_buffer);
+		instruction->saved_buffer = NULL;
+	}
+}
+
+static void save_view_buffer(struct sway_view *view,
+		struct sway_transaction_instruction *instruction) {
+	if (!sway_assert(instruction->saved_buffer == NULL,
+				"Didn't expect instruction to have a saved buffer already")) {
+		remove_saved_view_buffer(instruction);
+	}
+	if (view->surface && wlr_surface_has_buffer(view->surface)) {
+		wlr_buffer_ref(view->surface->buffer);
```

Comment:
What do we want to do with subsurfaces and popups?

Reply:
That's a good question. If the user has a popup open and presses a keybind then we should execute it (eg. screenshot), but maybe we should either reposition or close popups when we apply a transaction. I guess with subsurfaces we'd have to play on the safe side and keep them open.